Add new feed for .NET 5

ankitkumarr opened the pull request:
Once this is settled on, approved and merged, we would need to complete these tasks--

- [ ]  Update the build and release steps to have this auto-updated similar to our other feeds.
- [ ]  Add the new feed to the storage account / CDN.

ankitkumarr commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,317 @@
+      "displayVersion": "v1",
+      "releaseQuality": "GA",
+      "targetFramework": ".NET Framework",
+      "capabilities": "",
```
Added `"capabilites"` that we can use in future. Tools would be expected to only iterate over tags/trains that they have the capability for.

ankitkumarr commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,317 @@
+    },
+    "v3-isolated": {
+      "release": "3.19.0-isolated",
+      "displayName": ".NET 5 (isolated)",
```
All display names are modified to reflect the .NET version.

ankitkumarr commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,317 @@
+      ],
+      "serviceSiteConfig": [
+        "netFrameworkVersion"
+      ],
```
I have added the app settings and site config updates required to create/publish to a function app in Azure. And the properties `"serviceAppSettings"` and `"serviceSiteConfig"` that needs to be looked at for a given release version.

ankitkumarr commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,317 @@
+      "nodeVersion": "~12",
+      "localEntryPoint": "func.exe",
+      "itemTemplates": "https://ankikumapublic.blob.core.windows.net/templates/Microsoft.Azure.WebJobs.ItemTemplates.4.1.1675.nupkg",
+      "projectTemplates": "https://ankikumapublic.blob.core.windows.net/templates/Microsoft.Azure.WebJobs.ProjectTemplates.4.1.1675.nupkg",
```
These are the links to the templates that contain some of the new set of templates (WIP) required for this isolated release. We are thinking that we will keep the template version for isolated and in-proc separate for ease of maintenance and management.

I am even open to consider a different package altogether. But for now, just separating via major versions should be fine.

ankitkumarr commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,317 @@
+      "FUNCTIONS_WORKER_RUNTIME": "dotnet-isolated",
+      "netFrameworkVersion": "v5.0",
+      "requiredRuntime": ".NET",
+      "minimumRuntimeVersion": "5.0",
```
We need to make sure the `"requiredRuntime"` and `"minimumRuntimeVersion"` are in the format expected by the tools.

ankitkumarr commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,317 @@
+      "capabilities": "",
+      "hidden": false
+    },
+    "v3-isolated": {
```
We need to confirm that adding the `-isolated` is acceptable by everyone. I would imagine that because we are going with a new file completely, this would be ok. If right now tools expect a certain format, they can be updated as we only expect net5 to work from future release of the tools. Past releases would continue to use the old feed.

soninaren commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,239 @@
+      "localEntryPoint": "func.exe",
+      "templateApiZip": "https://functionscdn.azureedge.net/public/TemplatesApi/2.1.1642.zip",
+      "extensionBundle": {
+        "hostConfig": {
```
remove `hostConfig`. No one is using it.

ankitkumarr replied:
Is `extensionBundle` even used by anyone?

soninaren replied:
no, sorry i mean to say remove `extensionBundle`

soninaren commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,239 @@
+          "sha2": "f8cff611d0cc2198589a7ff0b1b268549bf86c11afd799098df59f7721d6b771"
+        },
+        {
+          "OperatingSystem": "MacOS",
```
Change to OS, or change others to OperatingSystem.

soninaren commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,239 @@
+      }
+    },
+    "2.60.0": {
+      "cli": "https://functionscdn.azureedge.net/public/2.7.3188/Azure.Functions.Cli.min.win-x86.2.7.3188.zip",
```
remove this and add to the standalone section, with a boolean value `"Default" : true`

phil-allen-msft replied:
It also would need a different operating system or some other distinguisher to make it different from the Windows/x86 entry already in the standalone cli set.

ankitkumarr replied:
I think the difference is that this is `minified`. Maybe I will move there and add an entry saying `"size" : "minified"` and others would have `"size" : "full"`? Unless there's a better term to encalpsulate that -- maybe `type` or `compression`.

soninaren commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,239 @@
+      "cli": "https://functionscdn.azureedge.net/public/2.7.3188/Azure.Functions.Cli.min.win-x86.2.7.3188.zip",
+      "sha2": "f04accdc23432695786f434835b12558e603866a5c3b6d5c398965d3dccc78e9",
+      "localEntryPoint": "func.exe",
+      "templateApiZip": "https://functionscdn.azureedge.net/public/TemplatesApi/2.1.1642.zip",
```
Rename `templateApiZip` to  `StaticTemplates` or `templates`

ankitkumarr replied:
That makes sense. I tried to avoid rename changes (from the last feed) unless necessary. But I am guessing because the structure has changed, might as well rename things as appropriate. I will make these updates.

soninaren commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,239 @@
+    "2.60.0": {
+      "cli": "https://functionscdn.azureedge.net/public/2.7.3188/Azure.Functions.Cli.min.win-x86.2.7.3188.zip",
+      "sha2": "f04accdc23432695786f434835b12558e603866a5c3b6d5c398965d3dccc78e9",
+      "localEntryPoint": "func.exe",
```
this and sha2 should be moved in the CLI section as well.

phil-allen-msft replied:
localEntryPoint is part of the debugging info and belongs in the runtimeInformation object.

soninaren commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,239 @@
+          }
+        }
+      },
+      "standaloneCli": [
```
Can rename to coreTools

soninaren commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,239 @@
+          }
+        }
+      },
+      "runtimeInformation": {
```
This is actually workerInformation

ankitkumarr replied:
Called it `workerRuntimeInformation` to map to our vocabulary of Worker runtime.

phil-allen-msft commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,239 @@
+            "sdkVersion": "3.0.1" ,
+            "itemTemplates": "https://www.nuget.org/api/v2/package/Microsoft.Azure.WebJobs.ItemTemplates/3.1.1648",
+            "projectTemplates": "https://www.nuget.org/api/v2/package/Microsoft.Azure.WebJobs.ProjectTemplates/3.1.1648",
+            "projectTemplateId": "Microsoft.AzureFunctions.ProjectTemplate.CSharp.3.x",
```
This is a C# specific value; how about "projectTemplateId": { "csharp" : "the value" }? There are other dotnet languages that would have different templates.

phil-allen-msft commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,239 @@
+            "projectTemplateId": "Microsoft.AzureFunctions.ProjectTemplate.CSharp.1.x",
+            "serviceAppSettings": {
+              "FUNCTIONS_EXTENSION_VERSION": "~1"
+            },
```
Let's add          "FUNCTIONS_WORKER_RUNTIME": null here, so that if someone (not sure why) were to publish a v1 project to a function app that was on v2, we would clear the property.  Scenario-wise it's not super necessary, but testing-wise it proves we handle clearing properties.

ankitkumarr replied:
@phile-allen-msft, I have added `FUNCTIONS_WORKER_RUNTIME: dotnet` as technically that's a valid setting and is usually just assumed if not there for v1 I think.

phil-allen-msft commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,239 @@
+              "diplayVersion": "v3",
+              "targetFramework": ".NET 5"
+            },
+            "capabilites": "outOfProc",
```
let's add a net5 capability with the comma separator, to document how we expect mutliple entries to work.

ejizba commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,269 @@
+              "displayName": ".NET 5 (Isolated)",
+              "hidden": true,
+              "diplayVersion": "v3",
+              "targetFramework": ".NET 5"
```
Is "targetFramework" used? I would think "displayName" is better for UI and "capabilities" (or something outside "displayInfo") is better for determining framework

ankitkumarr replied:
@phil-allen-msft, do you know if/how `targetFramework` is used today?

phil-allen-msft replied:
Compiling without it.... it's the branded (not localized) part of the display:
![image](https://user-images.githubusercontent.com/32652734/109064065-c52e7a80-769e-11eb-90b5-52fcb318af3a.png)

Similar to Eric's picture above.

ankitkumarr replied:
If the displayName has that now, do we still need that?

ejizba commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,269 @@
+        "dotnet": {
+          "net5-isolated": {
+            "displayInfo": {
+              "displayName": ".NET 5 (Isolated)",
```
[Stacks API](https://github.com/Azure/azure-functions-ux/blob/dev/server/src/stacks/2020-10-01/stacks/function-app-stacks/Dotnet.ts#L9) uses `.NET 5 (non-LTS)`. Would be nice to pick one and be consistent

phil-allen-msft replied:
The display name can only have non-translatable parts.  Anything that needs to be displayed with translation will need some boolean/enum properties that tell the client how to assemble the name in a translated form.  ".NET 5" is branding, no translation.  "Isolated" and "non-LTS" both require translation.

ejizba replied:
What if we have a separate "description" property for "Isolated"/"non-LTS". It doesn't solve localizability, but I imagine it would give us more flexibility in the future. Also, VS Code has an established pattern for the UI of "secondary" information, meaning we would do something like this:
<img width="231" alt="Screen Shot 2021-02-24 at 12 01 40 PM" src="https://user-images.githubusercontent.com/11282622/109058679-24d55780-7698-11eb-9ed0-a364a1608c77.png">

ankitkumarr replied:
Ok, I will add a `description` property unless opposed.

ejizba commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,269 @@
+              "diplayVersion": "v3",
+              "targetFramework": ".NET 5"
+            },
+            "capabilites": "outOfProc,net5",
```
Can we be consistent and use "isolated" instead of "outOfProc" in all cases?

Also, I think it's fair to assume it's "inProc" unless "isolated" is specified, so I don't know that we even need to specify "inProc" anywhere. Similar to how `FUNCTIONS_WORKER_RUNTIME` is `dotnet` vs. `dotnet-isolated`, not `dotnet-inProc` vs. `dotnet-outOfProc`

ankitkumarr replied:
Yeah that makes sense.

ejizba commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,269 @@
+  "releases": {
+    "1.13.0": {
+      "templates": "https://functionscdn.azureedge.net/public/TemplatesApi/1.0.3.10182.zip",
+      "workerRuntimeInformation": {
```
nit: Can we do `workerRuntimes` to match `tags` and `releases` (as opposed to `tagInformation` and `releaseInformation`)

ejizba reviewed (approved):
A few comments, but I think we could get this to work as-is in VS Code. Tbh, we don't use nearly as many properties as VS does (mostly because we don't have to worry about long-term support of older versions of tooling) - so it's probably a bit more important for them.

ankitkumarr commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,269 @@
+              "FUNCTIONS_EXTENSION_VERSION": "~3",
+              "FUNCTIONS_WORKER_RUNTIME": "dotnet-isolated"
+            },
+            "windowsSiteConfig": {
```
yeah @ejizba, that will always be updated for all languages. This one is very specific to VS (.NET only). You should be fine ignoring these on your end.

ankitkumarr commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,269 @@
+            "displayInfo": {
+              "displayName": ".NET 5 (Isolated)",
+              "hidden": true,
+              "diplayVersion": "v3",
```
I kept this similar to how we had it in the last feed. If it's unused, yeah we can remove this. @phil-allen-msft, do you we can safely remove this property?

ejizba replied:
Yeah it was in the last feed, but it was on the _tag_. "v3" is related to the func version, not the .net version

phil-allen-msft replied:
I believe that the display version was interesting when there are 'prerelease' tags to also display.  v3-preview, for instance, could still need to display as 'v3' in some cases.

ankitkumarr replied:
I think we can leave it there as the `prerelease` tags are still relevant though. @ejizba, yeah I think that refers to the Functions versions, which is still V3.

ejizba commented on cli-feed-v3-3.json:
```diff
@@ -0,0 +1,269 @@
+              "diplayVersion": "v3",
+              "targetFramework": ".NET 5"
+            },
+            "capabilites": "outOfProc,net5",
```
Can we add a targetFramework property that maps to the [monikor](https://docs.microsoft.com/en-us/dotnet/standard/frameworks) in the user's csproj (aka net5, netcoreapp3.1)?

I imagine we could get into a situation where there are multiple worker runtimes like this:

- net8 isolated
- net8 inproc
- net9 isolated
- net9 inproc

I can use the sdk name from the csproj to handle inproc vs. isolated, but I need something to handle net8 vs net9. I don't want to use the existing `displayInfo.targetFramework` because that's for display and I'd still have to map it to the monikor. I suppose we could do something with "capabilities", but I'd prefer a separate targetFramework property.

phil-allen-msft replied:
That makes a lot of sense as a sibling to sdkName/localEntryPoint.

ankitkumarr commented:
@ejizba and @phil-allen-msft, I think at this point I should have resolved all / most of the comments. We could iterate over for the display information later, but in the interest to get an intial version released, I will merge this. However, please continue using this PR for more discussion, and I can open another issue / PR to update things as we need.

In my most recent change, I have:
- renamed this file to be `cli-feed-v4.json`
- added proper links for templates 
- set hidden property to false for net5
- moved `sdkName` and `sdkVersion` to be under `sdk: {}`.

phenning commented:
@ankitkumarr I just noticed today that "capabilites" has a misspelling in the file while I was working on updating our model code in tooling. It should be capabilities
